### ISSUE TITLE
Add tonemapping filter to convert HDR10 content to SDR.

### ISSUE THREAD

**Issue by EraYaN** [jellyfin/jellyfin#415]
This for clients that are SDR only. Now the colors are all washed out. Newer ffmpeg has a pretty nice tonemapping filter.

**Comment by sparky8251** [comment]
Since this involves moving up to ffmpeg 4.1.x from 4.0.x its going to be a long process. That said, I agree its something we should do.

**Comment by gerrit507** [comment]
What your trying to achieve is possible with the zscale paramater, which is supported since ffmpeg 3.0.

Could you please explain what new filter in ffmpeg 4.1 you're talking about?

About zscale:
https://stevens.li/guides/video/converting-hdr-to-sdr-with-ffmpeg/

**Comment by EraYaN** [comment]
Mmm I'm not sure but does zscale use the metadata? Since it might clip into white I believe, and some time ago the tonemapping was still single core in ffmpeg.

**Comment by Essasetic** [comment]
Since FFMpeg 4.2 got implemented. Is this possible now?

**Comment by EraYaN** [comment]
Should be, although some of my testing showed it's still dark magic. Some of the numbers required are complicated domain specific knowledge. On my home desktop I have some test command lines, with output comparisons. I'll see if I can VPN into home and get them.

I also did some HDR transcoding tests (as in HDR output) but that is even more magic, lots of numeric parameters.

**Comment by ggogel** [comment]
Just for your information, I use the following ffmpeg parameters for HDR -> SDR tonemapping:
zscale=t=linear:npl=100,format=gbrpf32le,zscale=p=bt709,tonemap=tonemap=hable:desat=0,zscale=t=bt709:m=bt709:r=tv,format=yuv420p

**Comment by camjac251** [comment]
There's various documentation available for the filters. Tonemapping is done through a filter
It might be good to use mediainfo with ffmpeg to classify files into what colorspace they are first in because ffmpeg might need an input colorspace before it can convert.
Tonemapping has VAPPI and OpenCL HW Accelerated options as well.

There's also various algorithms available for tonemapping, which would be nice if the user could choose: clip, linear, gamma, reinhard, hable, mobius,

https://ffmpeg.org/ffmpeg-filters.html#tonemap-1
https://ffmpeg.org/ffmpeg-filters.html#tonemap_005fvappi
https://ffmpeg.org/ffmpeg-filters.html#tonemap_005fopencl

 It's also available here https://ffmpeg.org/doxygen/trunk/structTonemapContext.html but it might be easier to create or find a wrapper for .NET that has the filters defined already.

@EraYaN What testing have you performed? Could you elaborate on the

> complicated domain specific knowledge

This feature shouldn't be too complex if a wrapper is used for the in between of Jellyfin and ffmpeg. FFMpeg already takes care of the majority of the calculations, it just needs a few parameters for the effect.

To add to ggogel's parameters:
Running ffmpeg with the following arguments can tonemap a 2160p HDR10 Bluray with ease.
`ffmpeg-i HDRBlu-ray.mkv -vf zscale=t=linear:npl=100,format=gbrpf32le,zscale=p=bt709,tonemap=tonemap=hable:desat=0,zscale=t=bt709:m=bt709:r=tv,format=yuv420p -c:v libx264 -crf 17 -preset slower TonemappedFile.mkv`

This converts it down from Rec2020/Rec2100 to Rec709 for an SDR screen to view. It could also help with thumbnails or chapter images as that has no proper colorspace conversion either.

**Comment by EraYaN** [comment]
Alright I found my old command lines:

```bat
ffmpeg -y -i test.mkv -vf select=gte(n\,140) -vframes 1 output/output-areference.png

ffmpeg -y -i test.mkv -vf select=gte(n\,140),zscale=t=linear:npl=100,format=gbrpf32le,zscale=p=bt2020,tonemap=tonemap=hable,zscale=t=bt709:m=bt709:r=tv,format=yuv420p -vframes 1 output/output-stevenli-bt2020.png

ffmpeg -y -i test.mkv -vf select=gte(n\,140),zscale=t=linear:npl=100,format=gbrpf32le,zscale=p=bt2020,tonemap=tonemap=hable:desat=0,zscale=t=bt709:m=bt709:r=tv,format=yuv420p -vframes 1 output/output-stevenli-bt2020-desat0.png

ffmpeg -y -i test.mkv -vf select=gte(n\,140),zscale=t=linear:npl=100,format=gbrpf32le,zscale=p=bt709,tonemap=tonemap=hable,zscale=t=bt709:m=bt709:r=tv,format=yuv420p -vframes 1 output/output-stevenli.png

ffmpeg -y -i test.mkv -vf select=gte(n\,140),zscale=t=linear:npl=100,format=gbrpf32le,zscale=p=bt709,tonemap=tonemap=hable:desat=0,zscale=t=bt709:m=bt709:r=tv,format=yuv420p -vframes 1 output/output-stevenli-desat0.png

ffmpeg -y -i test.mkv -vf select=gte(n\,140),zscale=transfer=linear,tonemap=tonemap=clip:param=1.0:desat=2:peak=0,zscale=transfer=bt709,format=yuv420p -vframes 1 output/output-doom9-1.png

ffmpeg -y -i test.mkv -vf select=gte(n\,140),zscale=tin=smpte2084:min=bt2020nc:pin=bt2020:rin=tv:t=smpte2084:m=bt2020nc:p=bt2020:r=tv,zscale=t=linear:npl=100,format=gbrpf32le,zscale=p=bt709,tonemap=tonemap=hable:desat=0,zscale=t=bt709:m=bt709:r=tv,format=yuv420p -vframes 1 output/output-doom9-2.png
```

I like output-doom9-2.png and output-stevenli.png and output-stevenli-desat0.png (in order)
they seem to be the closest to what madVR does. Comparing on my LG TV is a bit harder but they seem fairly close in tone.

This is for HDR10 content (bt2020)

**Comment by camjac251** [comment]
To add to EraYaN's HDR -> HDR conversions, it will likely be necessary to add metadata to the file in order for a player to see it being HDR10 again
`ffmpeg -i HDRBlu-ray.mkv -c:v libx265 -x265-params "level=5.2:colorprim=bt2020:colormatrix=bt2020nc:transfer=smpte2084" -c:a copy -bsf:v hevc_metadata=colour_primaries=9:transfer_characteristics=16:video_full_range_flag=-1:matrix_coefficients=9 HDR-Encode.mkv`

**Comment by EraYaN** [comment]
The bottom one from my omment seems to trigger the HDR playback on my LG TV (webOS 4.0) But yes it probably needs coefficients from the source file (luminance etc) too.

**Comment by Z3ROGRAV1TYx** [comment]
+1 for this implementation, hdr content isn't watchable on non-hdr displays due to colors

**Comment by Shadowghost** [comment]
Any news on this? Would be a nice addition to have this working 😄

**Comment by dkanada** [comment]
Haven't seen any news on this front lately.

**Comment by MrTimscampi** [comment]
On the client side I have noticed that Video tracks have a VideoRange property.

This seems to always be set on SDR, even for HDR10 or Dolby Vision movies.

Here are the results of `ffmpeg -i` for two files.

**HDR10**

```
Stream #0:0: Video: hevc (Main 10), yuv420p10le(tv, bt2020nc/bt2020/smpte2084), 3840x2160 [SAR 1:1 DAR 16:9], 23.98 fps, 23.98 tbr, 1k tbn, 23.98 tbc (default)
    Metadata:
      BPS             : 58958985
      BPS-eng         : 58958985
      DURATION        : 01:48:19.493000000
      DURATION-eng    : 01:48:19.493000000
      NUMBER_OF_FRAMES: 155832
      NUMBER_OF_FRAMES-eng: 155832
      NUMBER_OF_BYTES : 47900439030
      NUMBER_OF_BYTES-eng: 47900439030
      _STATISTICS_WRITING_APP: mkvmerge v18.0.0 ('Apricity') 64-bit
      _STATISTICS_WRITING_APP-eng: mkvmerge v18.0.0 ('Apricity') 64-bit
      _STATISTICS_WRITING_DATE_UTC: 2017-12-13 20:50:29
      _STATISTICS_WRITING_DATE_UTC-eng: 2017-12-13 20:50:29
      _STATISTICS_TAGS: BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES
      _STATISTICS_TAGS-eng: BPS DURATION NUMBER_OF_FRAMES NUMBER_OF_BYTES
```

**Dolby Vision**

```
Stream #0:0(und): Video: hevc (Main 10) (hev1 / 0x31766568), yuv420p10le(tv, bt2020nc/bt2020/smpte2084), 3840x2160 [SAR 1:1 DAR 16:9], 50190 kb/s, 23.98 fps, 23.98 tbr, 24k tbn, 23.98 tbc
    Metadata:
      creation_time   : 2019-12-21T17:54:12.000000Z
      handler_name    : video handler
      encoder         : HEVC Coding
```

**Comment by so-rose** [comment]
Well I didn't like that my stuff is unwatchable (or, well, dull-looking), so I started playing around. That's when I found this thread.

That was a week ago :sweat_smile: 

But - I solved it! Scroll down :smile: 

# Simple Solution: The Easy Way
For a quick fix, doing a network stream (using "Copy Stream URL") through VLC performs perfect gamut and tone mapping (it seems to detect the source gamut and HDR status through the file metadata).

*With the caveat that you have to stream the 4K file unaltered. I'm lucky in that my wifi seems to be able to handle that, kind of, but even if it could, for some reason my poor Chromebook has a lot of trouble decoding 10-bit 4K HEVC in real time.*

*It doesn't fix the web interface or non-native Kodi client either. So it's not really a solution.*

# The Hard Way: Can we Hack It?

**Tl;dr** Yes! I present to you, a horrible time sink - https://github.com/so-rose/fast-hdr

It's a Python wrapper for `ffmpeg`, which intercepts, builds, and runs an identical (real) ffmpeg command, with one difference: The video stream has a pipe in the middle. It pipes through a custom, multithreaded HDR --> SDR conversion program written in `C++`.

## The Journey: Why Not Modify the Jellyfin Source Code?

Because that's boring!

## The Journey: The Idea

It occurred to me that we can change the way transcoding happens, without touching the Jellyfin source code, simply by pointing Jellyfin at an alternative ffmpeg "binary" (=a script that calls FFMPEG with some modifications! )

## The Journey: Injecting Code Between FFMPEG and Jellyfin

I created the following super-hacky "ffmpeg binary", which essentially hijacks the FFMPEG command sent by Jellyfin, and injects the desired video filter option.

This is what I pointed Jellyfin at:
```python3
import os, sys

ffmpg = '/usr/lib/jellyfin-ffmpeg/ffmpeg'
filt = "zscale=t=linear:npl=100,format=gbrpf32le,zscale=p=bt709,tonemap=tonemap=hable:desat=0,zscale=t=bt709:m=bt709:r=tv,format=yuv420p"

x = sys.argv[1:]
for i in [3, 35, 43, 56, -7, -1] :
        if i == 43 :
                x[i] = '"' + x[i] + "," + filt + '"'
        else :
                x[i] = '"' + x[i] + '"'


cmd = ' '.join( [ffmpg] + x )

os.system(cmd)
```

After setting Jellyfin to use `ffmpeg.py` as its `ffmpeg` binary, I noticed it was performing the following command:
```
<PATHTO>/ffmpeg.py -f matroska,webm -i file:"<IN_FILE>.mkv" -map_metadata -1 -map_chapters -1 -threads 0 -map 0:0 -map 0:5 -map -0:s -codec:v:0 libx264 -pix_fmt yuv420p -preset veryfast -crf 23 -maxrate 34541128 -bufsize 69082256 -profile:v high -level 4.1 -x264opts:0 subme=0:me_range=4:rc_lookahead=10:me=dia:no_chroma_me:8x8dct=0:partitions=none  -force_key_frames:0 "expr:gte(t,0+n_forced*3)" -g 72 -keyint_min 72 -sc_threshold 0 -vf "scale=trunc(min(max(iw\,ih*dar)\,1920)/2)*2:trunc(ow/dar/2)*2" -start_at_zero -vsync -1 -codec:a:0 aac -strict experimental -ac 2 -ab 384000 -af "volume=2" -copyts -avoid_negative_ts disabled -f hls -max_delay 5000000 -hls_time 3 -individual_header_trailer 0 -hls_segment_type mpegts -start_number 0 -hls_segment_filename "<HLS_SEG>" -hls_playlist_type vod -hls_list_size 0 -y "<OUT_FILE>"
```

## The Journey: It Works! Kind of.

It worked!

Well - there were some problems.
- The hard-coded argument indices broke seeking, but this could be fixed with smarter index-finding.
- It was *really, really slow* (on my 32-thread server).

But, alas, there are other options, no? I tried baking the PQ+BT.2020 to Rec.709+BT.709 transform into a hard-coded 3D LUT, generated by [LUTcalc](https://cameramanben.github.io/LUTCalc/):

```python3
filt = 'lut3d=' + '/data/main/jellyfin/bt2020_bt709.cube'
```

Faster, but still too slow. There's a reason LUT processing tends to happen on GPUs (of which my server has none).

Also, transfer functions with 3D LUTs isn't very good (and `ffmpeg` has no 1D LUT support).  Not to mention, ffmpeg only support up to 64x64x64 3D LUTs.

## The Journey: So, it Can't Be Done?

`ffmpeg` clearly can't do this alone with sufficient speed. But we maybe we can solve it anyway :)

Let's run with the same concept of injecting a custom `ffmpeg` call based on Jellyfin's call. First, we mine some key data using `mediainfo`; stuff like seek time, the file path, source colorspace, transfer properties, and resolution.

Then, here's the idea:
1. Let ffmpeg read the file, YUV data and spit it out to `stdout`.
2. Let some super simple, purpose-build program execute an HDR --> SDR pipeline, taking raw YUV data from `stdin` and pushing raw YUV data to `stdout`.
3. Let `ffmpeg` read the converted raw YUV data and continue processing it the way Jellyfin desires.

In pipe form, I arrived at the following pseudo-solution:
```bash
ffmpeg -ss "$TIME" -i file:"$FILE" -c:v rawvideo -pix_fmt yuv420p -an -f rawvideo pipe:1 | <HDR_TO_SDR> | ffmpeg -s "$RES" -pix_fmt yuv420p -f rawvideo -i pipe:0 -ss "$TIME" -i file:"$FILE" -map 0:v -map 1:a <JELLYFIN_ENCODING_STUFF>
```

First, we test that `ffmpeg` can play video smoothly in this way:
```bash
ffmpeg -ss "$TIME" -i file:"$FILE" -threads 0 -c:v rawvideo -pix_fmt yuv420p -an -f rawvideo pipe:1 | ffmpeg -s 3840x2160 -pix_fmt yuv420p -f rawvideo -i pipe:0 -ss "$TIME" -i file:"$FILE" -threads 0 -map 0:v -map 1:a -c:v libx264 -pix_fmt yuv420p -preset veryfast -crf 23 -maxrate 34541128 -bufsize 69082256 -level 4.1 -profile:v high -c:a aac -f avi pipe:1 | ffplay pipe:0
```

This piping stuff does not seem to bottleneck `ffmpeg`; a 4K/10bit HDR HEVC video plays perfectly.

Now, all we need is a CPU-based HDR to SDR conversion that can read (from stdin), process, and write (to stdout) a 4K YUV frame in 40ms (for a min. framerate of ~24FPS, so that it doesn't bottleneck). Easy...s

## The Journey: Great, But - What Kind of Processing??

It [seemed to me as though (page 19-20)](https://www.itu.int/dms_pub/itu-r/opb/rep/R-REP-BT.2390-3-2017-PDF-E.pdf) (also after reverse engineering by comparing results with VLC) we needed the following processing pipeline:
1. YUV --> RGB
2. Linearization (Inverse PQ/HLG Curve)
3. Tonemapping/EETF (Reinhard is a good starting point, Hable is default in VLC)
4. Gamma Correction (Rec.709 Curve)
5. RGB --> YUV

Why no BT.2020-->Rec.709 gamut? Because it would seem that `ffmpeg` already does this! I get why `ffmpeg` doesn't invert the PQ curve for me; `ffmpeg` would then need to make a creative tone-mapping choice before it's anywhere near what 8 bits can hold.

In any case, this saves us having to deal with anything out-of-gamut.

BUT. It ended up being too slow, no matter how many `#pragma omp`'s I littered around.

## The Journey: Are LUTs Always the Answer?
I took the entire YUV444p -- (stuff) --> YUV444p and performed it with `double` floating point precision, then encoded it into a `3 * 256^3` sized LUT. In other words, **every single 8-bit value has had the entire imaging pipeline performed, and can be indexed**.

After a bunch of multithreading of the `STDIN` reader / `STDOUT` writer, 

With a LUT lookup being the only math happening, I ended up with `30` FPS on my Threadripper 1950x at 4K, with over `80` at 1080p.


# The End

After a lot of code, hours of optimization, and a surprising amount of reverse engineering `ffmpeg` source code, I present to you - `fast-hdr`!

https://github.com/so-rose/fast-hdr

What can I say - it works :)

Kind of. Don't try to seek. It's not fast...

**Comment by Window-Hero** [comment]
So, from what I'm able to gather, the bottleneck was that Jellyfin needed to update to ffmpeg 4.1.x, but now Jellyfin is on ffmpeg 4.2.1, so is tonemapping still in the works or what?  Is it implemented and there's some magic setting I just don't know about?  And a little bit off topic, but what about for displays that do support HDR content?  Is there any way to get actual HDR working on those other than copying the stream link into an external player?

**Comment by EraYaN** [comment]
HDR should work out of the box (for Direct Streaming and Direct Play) for all clients that support it (including Kodi on Android TV).
It's transcoding that is the issue, and there is no PR yet as far as I know.

**Comment by Window-Hero** [comment]
I understand that direct play works if you copy the stream link into something like VLC (though I've had no luck actually getting HDR working in VLC in general, I think that's just because my monitor's HDR support is garbage), but all of my HDR content (and I assume this is common) is HEVC, which is not widely supported by web browsers, which forces Jellyfin to transcode it to SDR, which it then does poorly, despite having theoretically the correct version of ffmpeg required to do the tonemapping it needs to do.

**Comment by MrTimscampi** [comment]
> I understand that direct play works if you copy the stream link into something like VLC (though I've had no luck actually getting HDR working in VLC in general, I think that's just because my monitor's HDR support is garbage), but all of my HDR content (and I assume this is common) is HEVC, which is not widely supported by web browsers, which forces Jellyfin to transcode it to SDR, which it then does poorly, despite having theoretically the correct version of ffmpeg required to do the tonemapping it needs to do.

There is no way to support HDR in browsers currently with Jellyfin.

We would need HDR10 to HLG conversion, which is a whole different thing than normal tone mapping and would still require transcoding to VP9 or H.264.

If you want to play HDR content on a PC right now, I'd suggest using Jellyfin MPV Shim.

**Comment by Window-Hero** [comment]
Okay, that's nice, because it streamlines viewing HDR content, but what is the status of tonemapping, so that if you do try watching in browser, it doesn't feel like you've suddenly gone colorblind?

**Comment by MrTimscampi** [comment]
You can follow progress on #3442

**Comment by reacocard** [comment]
would is be possible to add support for doing the cpu tonemap with forced downscaling? my 8-core zen2 server can't keep up with 4k cpu tonemap but it can do 4k downscaled to 1080p tonemap at ~1.5x realtime, and being able to stream 4k HDR files to SDR devices at 1080p would be better than not being able to stream them at all.

ffmpeg command i used to test this: `fmpeg -i 'infile.mkv' -vf scale=1920x1080:force_original_aspect_ratio=decrease,zscale=transfer=linear,tonemap=tonemap=mobius:param=0.2,zscale=transfer=bt709,format=yuv420p -c:v libx264 -preset fast -tune film -maxrate 50M -c:a copy -c:s copy -c:d copy -c:t copy "outfile.mkv"

**Comment by nyanmisaka** [comment]
Most modern 8-core cpu can handle tonemapping from 4k HDR to 1080p SDR 25fps with `tonemap` filter.
But it will be a nightmare for streaming when it comes to 4k or 60fps tonemapping and encoding.

This is feasible for self-build server with a powerful cpu but not for those pre-built NAS such as Synology.
I would like to use `tonemap_vaapi` on these low voltage chips but it is still pending as that filter isn't working as expected on intel Gen9 chips.

From my test, Vega 8 APU with `tonemap_opencl` filter brings me up to ~40fps on 4k tonemaping and ~80fps on 1080p.

**Comment by tristan-k** [comment]
> This is feasible for self-build server with a powerful cpu but not for those pre-built NAS such as Synology.
> 
> I would like to use `tonemap_vaapi` on these low voltage chips but it is still pending as that filter isn't working as expected on intel Gen9 chips.

Does [https://github.com/jellyfin/jellyfin-ffmpeg/pull/59](#59) solves the issue for hardware acceleration om intel quick sync version 5 (skylake) and version 6 (Kaby Lake, Coffee Lake, Whiskey Lake, Comet Lake)?

**Comment by cvium** [comment]
This is implemented in 10.7. Any issues regarding this should be opened as separate issues.
